File: common/src/main/java/dev/gallon/motorassistance/common/event/SingleEventBus.java
package dev.gallon.motorassistance.common.event;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SingleEventBus {
    public static SingleEventBus INSTANCE = new SingleEventBus();

    static Map<Class<? extends Event>, List<Callback<? extends Event>>> callbacksByEvent = new HashMap<>();

    static public <T extends Event> void listen(Class<T> event, Callback<T> callback) {
        if (!callbacksByEvent.containsKey(event)) {
            callbacksByEvent.put(event, new ArrayList<>());
        }
        callbacksByEvent.get(event).add(callback);
    }

    static public <T extends Event> void publish(T event) {
        List<Callback<? extends Event>> callbacks = callbacksByEvent.get(event.getClass());
        if (callbacks != null) {
            callbacks.forEach(listener -> listener.callback(event));
        }
    }
}

File: fabric/src/main/kotlin/dev/gallon/motorassistance/fabric/events/SingleEventBus.kt
package dev.gallon.motorassistance.fabric.events

import kotlin.reflect.KClass

// It's probably not production-ready, but it works for the example
object SingleEventBus {

    val callbacks = mutableMapOf<KClass<out Event>, List<(event: Event) -> Unit>>()

    @Suppress("UNCHECKED_CAST")
    inline fun <reified T : Event> register(noinline callback: (event: T) -> Unit) {
        if (T::class in callbacks) {
            callbacks[T::class] = callbacks[T::class]!!.plus(callback as (event: Event) -> Unit)
        } else {
            callbacks[T::class] = listOf(callback as (event: Event) -> Unit)
        }
    }

    fun <T : Event> send(event: T) {
        callbacks[event::class]?.forEach { it.invoke(event) }
    }
}
